broker/cloud_functions/store_BigQuery/main.py

```diff
@@ -0,0 +1,155 @@
+    bq_dataset = f"{bq_dataset}_{TESTID}"
+    ps_topic = f"{ps_topic}-{TESTID}"
+
+schema_map = schema_maps.load_schema_map(SURVEY, TESTID)
+
+
+def run(msg: dict, context) -> None:
+    """Send alert data to various BigQuery tables.
+
+    Args:
+        msg: Pub/Sub message data and attributes.
+            `data` field contains the message data in a base64-encoded string.
+            `attributes` field contains the message's custom attributes in a dict.
+
+        context: Metadata describing the Cloud Function's trigging event.
+    """
+    # extract the alert
+    alert_dict = data_utils.decode_alert(
+        base64.b64decode(msg["data"]), drop_cutouts=True, schema_map=schema_map
+    )
+
+    # send the alert to BigQuery tables
+    a_table = insert_rows_alerts(alert_dict)
+    d_table = insert_rows_DIASource(alert_dict)
+
+    # announce what's been done
+    publish_pubsub(alert_dict, [a_table, d_table])
+
+
+def insert_rows_alerts(alert_dict):
+    """Insert rows into the `alerts` table via the streaming API."""
+    # send to bigquery
+    table_id = f"{bq_dataset}.alerts"
+    errors = gcp_utils.insert_rows_bigquery(table_id, [alert_dict])
+
+    # handle errors; if none, save table id for Pub/Sub message
+    if len(errors) == 0:
+        table_dict = {"alerts_table": f"{PROJECT_ID}:{table_id}"}
+    else:
+        msg = f"Error inserting to alerts table: {errors}"
+        logger.log_text(msg, severity="DEBUG")
+        table_dict = {"alerts_table": None}
+
+    return table_dict
+
+
+def insert_rows_DIASource(alert_dict):
+    """Insert rows into the `DIASource` table via the streaming API."""
+    table_id = f"{bq_dataset}.DIASource"
+
+    # extract the DIA source data
+    if SURVEY == "ztf":
+        source_dict = _extract_ztf_source(alert_dict)
+    elif SURVEY == "decat":
+        source_dict = _extract_decat_source(alert_dict)
+
+    # send to bigquery
+    errors = gcp_utils.insert_rows_bigquery(table_id, [source_dict])
+
+    # handle errors; if none, save table id for Pub/Sub message
+    if len(errors) == 0:
+        table_dict = {"DIASource_table": f"{PROJECT_ID}:{table_id}"}
+    else:
+        msg = f"Error inserting to DIASource table: {errors}"
+        logger.log_text(msg, severity="DEBUG")
+        table_dict = {"DIASource_table": None}
+
+    return table_dict
+
+
+def _extract_ztf_source(alert_dict):
+    # get candidate
+    dup_cols = ["candid"]  # candid is repeated, drop the one nested here
+    cand = {k: v for k, v in alert_dict["candidate"].items() if k not in dup_cols}
+
+    # get info for provenance
+    metakeys = ["schemavsn", "publisher", "objectId", "candid"]
+    metadict = {k: v for k, v in alert_dict.items() if k in metakeys}
+
+    # get string of previous candidates' candid, comma-separated
+    if alert_dict["prv_candidates"] is not None:
+        tmp = [pc["candid"] for pc in alert_dict["prv_candidates"]]
+        prv_candids = ",".join([f"{cid}" for cid in tmp if cid is not None])
```

Comment:
This line is a double loop: once for the list constructor and another for the join. Remove the brackets to only loop once: `prv_candids = ",".join(str(cid) for cid in tmp if cid is not None)`